### DISCUSSION TITLE
Introduce static modifier for a type parameter

### DISCUSSION THREAD

**Discussion by nvmkpk** [dotnet/csharplang#7363] [Language Ideas]
Currently a static type cannot be used as a type parameter in a generic class. Like below example:
```
static class Example1
{
}

class Generic1<T>
{
    public static void Sample1()
    {
        Generic1<Example1> x = null; // Produces error CS0718
    }
}
```

Proposal is to enable below:
```
static class Example1
{
}

class Generic1<static T>
{
    public static void Sample1()
    {
        Generic1<Example1> x = null;
    }
}
```

Adding static modifier would change the way it can be used within the generic class. A non-static type can still be used. Only things allowed with it would be to invoke static members or to use in `typeof` operator. Below will not be allowed:
```
class Generic1<static T>
{
    public T Value
    {
        get;
        set;
    }

}
```

This would be useful for cases like `ILogger<T>`. Currently we are forced to use `ILoggerFactory.CreateLogger(string)` method or  `ILoggerFactory.CreateLogger(Type)` extension method when we want to create logger for static type. The `static` modifier should be allowed to use together with variance modifiers in case of generic interfaces.

In case of `ILogger<>` its usage will be as shown below. This will allow both static and non-static types to be used as its type arguments.
```
public interface ILogger<static out TCategoryName> : ILogger
{
}
```

**Comment by theunrepentantgeek** [comment]
With support for **static** methods on interfaces, you can get very close with C# as it stands today; you might just need to create a non-static trampoline class to proxy. [[Sharplab](https://sharplab.io/#v2:D4AQTAjAsAULIGYAEBLAdgFwKYCcBmAhgMZZICSAYsRgPY4CeAPACoB8SA3rEj0okgQBGAZww5qfCADYkzJAGEcWAtgAUICAAYkaAgFssASgDcsAL6x4ycEgAKuYTTSduvfhu0A5faQ4BzLAxjJGFA4wsYXj5kexxHNHUtHR9DVx4uSKjebwMkAF5kg1NMngiIqz4wOwcnKiJaBiQALnI6hqZY+PZYDKj3aWq4pwUlFSxE7V0DQ3z2NCwAd0H41SmjYvKYdyqAWQJdAJxGCgAaWW6SpAWAC1xSCmbW6jomNh60nn45eSdRHABXeoTQpGJAfFyXLIgADsSAoADpFMo1GsTOCypYttYqvIIX1rAAWJB7dCqGYfXpZHgANwIOCQenyOkWxP2BEOjE6tWeDDOXLQrDJxSpNLpSAADky9IjfmJARhVAAiABCNEEipmIo+GJgQA===)]

**Reply by HaloFour** [reply]
I think it's worth reconsidering the imposed limitation that static classes cannot be type parameters, and that static classes cannot implement interfaces.  That would make this proposal unnecessary.

**Comment by ufcpp** [comment]
https://github.com/dotnet/csharplang/discussions/4840 ?

**Reply by nvmkpk** [reply]
This looks like solving same problem but with different syntax. That syntax feels little more confusing. With `where` clause it looks like only static type should be allowed.
